src/LoggerAnalyzer/LoggerAnalyzerAnalyzer.cs

```diff
@@ -33,42 +31,43 @@ public override void Initialize(AnalysisContext context)
-    }
-
-    private static void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
-    {
-        var constructorDeclarationSyntaxNode = (ConstructorDeclarationSyntax)context.Node;
-        foreach (var parameter in constructorDeclarationSyntaxNode.ParameterList.Parameters)
-        {
-            var genericParameterType = parameter.Type as GenericNameSyntax;
-            if (genericParameterType is null)
-            {
-                continue;
-            }
-
-            if (genericParameterType.TypeArgumentList.Arguments.Count != 1)
-            {
-                continue;
-            }
-
-            var parameterSymbol = context.SemanticModel.GetDeclaredSymbol(parameter, context.CancellationToken);
-            if (parameterSymbol.Type.Name != "ILogger")
-            {
-                continue;
-            }
-
-            var classSymbol = context.SemanticModel.GetDeclaredSymbol(constructorDeclarationSyntaxNode.Parent, context.CancellationToken);
-
-            var genericArgument = genericParameterType.TypeArgumentList.Arguments.First() as IdentifierNameSyntax;
-            var genericArgumentType = context.SemanticModel.GetTypeInfo(genericArgument, context.CancellationToken);
-            if (genericArgumentType.Type.Name.Equals(classSymbol.Name, StringComparison.OrdinalIgnoreCase))
-            {
-                continue;
-            }
-
-            context.ReportDiagnostic(Diagnostic.Create(Rule, genericArgument.GetLocation(), classSymbol.Name));
-        }
+
+        context.RegisterCompilationStartAction(context =>
+        {
+            var loggerSymbol = context.Compilation.GetTypeByMetadataName("Microsoft.Extensions.Logging.ILogger`1");
+            if (loggerSymbol == null)
+                return;
+
+            context.RegisterSymbolAction(context =>
+            {
+                var symbol = (IMethodSymbol)context.Symbol;
+                if (symbol.MethodKind != MethodKind.Constructor)
+                    return;
+
+                foreach (var parameter in symbol.Parameters)
+                {
+                    if (parameter.Type is not INamedTypeSymbol parameterType)
+                        continue;
+
+                    // Only consider ILogger<T> parameters
+                    if (!parameterType.OriginalDefinition.Equals(loggerSymbol, SymbolEqualityComparer.Default))
+                        continue;
+
+                    // Check <T> is the containing class
+                    if (parameterType.TypeArguments[0].Equals(symbol.ContainingType, SymbolEqualityComparer.Default))
+                        continue;
+
+                    // Try to report the type argument. If not possible, report the parameter
+                    var syntax = parameter.DeclaringSyntaxReferences[0].GetSyntax();
+                    if (syntax is ParameterSyntax { Type: GenericNameSyntax { TypeArgumentList.Arguments: var arguments } } && arguments.Count == 1)

```

Comment:
How can here be no argument if we checked the argument in the previous condition?

Reply:
It's hard to be sure the syntax is always the one you expect. C# provides multiple way to declare the same things. Also, you don't know what the next version will introduce. That's why I try to stay away from the syntax and use higher level APIs such as `IOperation` and `ISymbol`. I use the Syntax Apis only when there is no other way. 

Also, this code would work if you enable the support for VB.NET.